File: package/server/app/schemas/location.py
from datetime import date, datetime
from typing import Optional, List, Union
from uuid import UUID

from pydantic import BaseModel, Field
from app.schemas.photo import Photo

class LocationBase(BaseModel):
    name: str
    level: str
    count: int

class LocationStatistics(BaseModel):
    province_count: int
    city_count: int
    district_count: int
    country_count: int

class Location(LocationBase):
    id: Optional[str] = None
    is_custom: Optional[bool] = None
    cover: Optional[Photo] = None

    class Config:
        from_attributes = True

class MapMarker(BaseModel):
    id: str
    lat: float
    lng: float

class LocationValue(BaseModel):
    province: Optional[str] = None
    city: Optional[str] = None
    district: Optional[str] = None

class LocationSearchItem(BaseModel):
    label: str
    value: LocationValue

class TimelineNode(BaseModel):
    type: str = "default"
    startDate: str
    endDate: str
    locationName: str
    level: Optional[str] = None
    lat: Optional[float] = None
    lng: Optional[float] = None
    photoCount: int = 0
    coverId: Optional[UUID] = None
    startTime: Optional[datetime] = None
    endTime: Optional[datetime] = None
    
class TimelineResponse(BaseModel):
    nodes: List[TimelineNode]
    total: int

class TrajectoryPoint(BaseModel):
    photoId: UUID
    capturedAt: datetime
    endAt: Optional[datetime] = None
    lat: float
    lng: float
    photoCount: int = 1
    coverId: Optional[UUID] = None
    locationName: str
    level: str = "city"

class TrajectoryResponse(BaseModel):
    points: List[TrajectoryPoint]
    totalPhotos: int
    sampled: bool = False


class TimeCompareYear(BaseModel):
    year: int
    photo_count: int
    first_date: datetime
    last_date: datetime
    cover: Photo


class TimeCompareVisit(BaseModel):
    date: date
    photo_count: int
    first_time: datetime
    last_time: datetime
    cover: Photo


class TimeCompareSummary(BaseModel):
    eligible: bool
    reason: Optional[str] = None
    match_type: Optional[str] = None
    radius_m: Optional[int] = None
    visual_similarity: Optional[float] = None
    scene_id: Optional[UUID] = None
    location_name: Optional[str] = None
    location_address: Optional[str] = None
    city: Optional[str] = None
    source_photo_id: Optional[UUID] = None
    source_photo_year: Optional[int] = None
    years: List[TimeCompareYear] = Field(default_factory=list)
    visits: List[TimeCompareVisit] = Field(default_factory=list)
    first_photo: Optional[Photo] = None
    latest_photo: Optional[Photo] = None

File: package/website/src/types/location.ts
import type { Photo } from './album'

export interface Location {
  id?: string
  is_custom?: boolean
  name: string
  level: 'city' | 'province' | 'district' | 'scene'
  count: number
  cover: Photo | null
}

export interface LocationStatistics {
  province_count: number
  city_count: number
  district_count: number
  country_count: number
}

export interface Scene {
  id: string
  is_custom: boolean
  name: string
  description?: string
  level?: number
  address?: string
  latitude?: number
  longitude?: number
  radius?: number
  polygon?: number[][]
  photo_count?: number
  cover?: Photo | null
}

export interface SceneCreate {
  name: string
  description?: string
  level?: number
  address?: string
  latitude?: number
  longitude?: number
  radius?: number
  polygon?: number[][]
}

export interface SceneUpdate extends SceneCreate {}

export interface TimelineNode {
  type: string;
  startDate: string; // YYYY-MM-DD
  endDate: string; // YYYY-MM-DD
  locationName: string;
  level?: string; // Added level
  lat?: number;
  lng?: number;
  photoCount: number;
  coverId?: string;
  startTime?: string;
  endTime?: string;
}

export interface TimelineResponse {
  nodes: TimelineNode[];
  total: number;
}

export interface TimeCompareYear {
  year: number
  photo_count: number
  first_date: string
  last_date: string
  cover: Photo
}

export interface TimeCompareVisit {
  date: string
  photo_count: number
  first_time: string
  last_time: string
  cover: Photo
}

export interface TimeCompareSummary {
  eligible: boolean
  reason?: 'precise_location_required' | 'multiple_years_required' | string
  match_type?: 'scene' | 'nearby_gps' | string
  radius_m?: number
  visual_similarity?: number
  scene_id?: string
  location_name?: string
  location_address?: string
  city?: string
  source_photo_id?: string
  source_photo_year?: number
  years: TimeCompareYear[]
  visits: TimeCompareVisit[]
  first_photo?: Photo
  latest_photo?: Photo
}

export interface TrajectoryPoint {
  photoId: string;
  capturedAt: string;
  endAt?: string;
  lat: number;
  lng: number;
  photoCount: number;
  coverId?: string;
  locationName: string;
  level: string;
}

export interface TrajectoryResponse {
  points: TrajectoryPoint[];
  totalPhotos: number;
  sampled: boolean;
}
